rtCommon/bidsInterface.py

```diff
@@ -1,43 +1,292 @@
+        """
+        raise NotImplementedError('getNumVolumes not implemented for DicomBidsStream')
+
+    def getIncremental(self, volIdx) -> BidsIncremental:
+        """
+        Get the BIDS incremental for the corresponding DICOM image indicated
+            by the volIdx, where volIdx is equivalent to TR id.
+        Args:
+            volIdx: The volume index (or TR) within the run to retrieve.
+        Returns:
+            BidsIncremental for the matched DICOM for the run/volume
+        """
+        # wait for the dicom and create a bidsIncremental
+        dcmImg = self.dataInterface.getImageData(self.dicomStreamId, volIdx)
+        dicomMetadata = getDicomMetadata(dcmImg)
+        dicomMetadata.update(self.entities)
+        niftiImage = convertDicomImgToNifti(dcmImg)
+        incremental = BidsIncremental(niftiImage, dicomMetadata)
+        return incremental
+
+
+class BidsStream:
+    """
+    A class that opens a BIDS archive and prepares to stream the data as
+    BIDS incrementals.
+    """
+    def __init__(self, archivePath, **entities):
+        """
+        Args:
+            archivePath: Absolute path of the BIDS archive.
+            entities: BIDS entities (subject, session, task, run, suffix, datatype) that
+                define the particular subject/run of the data to stream
+        """
+        self.bidsArchive = BidsArchive(archivePath)
+        # TODO - when we have BidsRun
+        # self.bidsRun = self.bidsArchive.getBidsRun(**entities)
+        images = self.bidsArchive.getImages(**entities)
+        if len(images) == 0:
+            raise ValidationError('No matching images found')
+        if len(images) > 1:
+            raise ValidationError('More than one match, please give more specific subject/session/task/run')
+        self.bidsImage = images[0]
+        self.niftiImage = self.bidsImage.get_image()
+        self.filename = self.niftiImage.get_filename()
+        self.imgVolumes = nib.four_to_three(self.niftiImage)
+        self.metadata = self.bidsArchive.getSidecarMetadata(self.filename, includeEntities=True)
+        self.metadata.pop('extension')
+        self.numVolumes = len(self.imgVolumes)
+        self.nextVol = 0
+
+    def getNumVolumes(self) -> int:
+        """Return the number of brain volumes in the run"""
+        # TODO - when we have BidsRun
+        # return self.bidsRun.getNumVolumes()
+        return self.numVolumes
+
+    def getIncremental(self, volIdx) -> BidsIncremental:
+        """
+        Get a BIDS incremental for the indicated index in the current subject/run
+        Args:
+            volIdx: The volume index (or TR) within the run to retrieve.
+        Returns:
+            BidsIncremental of that volume index within this subject/run
+        """
+        # TODO - when we have BidsRun
+        # return self.bidsRun.getIncremental(volIdx)
+        if volIdx < 0:
+            volIdx = self.nextVol
+        if volIdx < self.numVolumes:
+            self.nextVol += 1
```

Comment:
I think clients may see some unexpected behavior here if they do a specific index after a sequence of negative indicies. E.g., client provides `volIdx` = -1, -1, -1, returning images 0, 1, and 2 and `nextVol = 3`. 

Now the client provides `volIdx` = 8, returning image 9, but also incrementing `nextVol` to 4. Thus, if the client continues providing -1, -1, -1 to resume the stream, they'll have missed an image. 

Perhaps the incrementing should go in the `if volIdx < 0` block?

Reply:
Perhaps the behavior should be that when the user specifies a volume it resets the nextVol counter to that position. So if the specify volIdx = 8, and then volIdx=-1, they will get the 9th and 10th images.

Reply:
I think that sounds like a good idea -- maybe add a comment on how that behavior works? (it sounds like we have an underlying pointer that is either moved to a particular index or just advanced, depending on if the # is >= 0 or < 0)